### PR TITLE
TASK-52372: XSS in create document form.

### PR DESCRIPTION
**Pull request by IlhemEssaadi**
ISSUES : XSS in create document form: Type a script in the document name when creating a new document, the script is executed and the alert popup displayed.
FIX : the problem fixed by adding a validator to the input text.

### CODE REVIEW

File: core/webui-explorer/src/main/java/org/exoplatform/ecm/webui/component/explorer/documents/UINewDocumentForm.java [hunk 1]
```diff
@@ -83,10 +86,11 @@
    * Constructor.
    *
    */
-  public UINewDocumentForm() {
+  public UINewDocumentForm() throws Exception{
```
Comment (rdenarie) [inline core/webui-explorer/src/main/java/org/exoplatform/ecm/webui/component/explorer/documents/UINewDocumentForm.java:87]:
```suggestion
  public UINewDocumentForm() {
```

File: core/webui-explorer/src/main/java/org/exoplatform/ecm/webui/component/explorer/documents/UINewDocumentForm.java [hunk 2]
```diff
@@ -83,10 +86,11 @@
...
     this.documentService = this.getApplicationComponent(DocumentService.class);
     // Title textbox
     UIFormStringInput titleTextBox = new UIFormStringInput(FIELD_TITLE_TEXT_BOX, FIELD_TITLE_TEXT_BOX, null);
+    titleTextBox.addValidator(MandatoryValidator.class).addValidator(ECMNameValidator.class).addValidator(XSSValidator.class);
```
Comment (rdenarie) [inline core/webui-explorer/src/main/java/org/exoplatform/ecm/webui/component/explorer/documents/UINewDocumentForm.java]: Why did you add ECMNameValidator ?

**Review by rdenarie — CHANGES_REQUESTED** [review]

File: core/webui-explorer/src/main/java/org/exoplatform/ecm/webui/component/explorer/documents/UINewDocumentForm.java (hunk 2 above)
Comment (IlhemEssaadi) [reply core/webui-explorer/src/main/java/org/exoplatform/ecm/webui/component/explorer/documents/UINewDocumentForm.java]: > Why did you add ECMNameValidator ?

The addition of the ECMNameValidator is mandatory so that the name of the document is compatible with jcr.

Comment (rdenarie) [reply core/webui-explorer/src/main/java/org/exoplatform/ecm/webui/component/explorer/documents/UINewDocumentForm.java]: In UINewDocumentForm, the title entered by the user is already updated to fit JCR constraint  :

>       title = getFileName(title, template);

So, no need to add a validator to check it.
The only needed validator is the one for XSS

**Review by rdenarie — CHANGES_REQUESTED** [review]

File: core/webui-explorer/src/main/java/org/exoplatform/ecm/webui/component/explorer/documents/UINewDocumentForm.java [hunk 3]
```diff
@@ -32,6 +32,8 @@
 
 import org.exoplatform.ecm.utils.text.Text;
 import org.exoplatform.ecm.webui.component.explorer.UIJCRExplorer;
+import org.exoplatform.ecm.webui.form.validator.ECMNameValidator;
```
Comment (rdenarie) [inline core/webui-explorer/src/main/java/org/exoplatform/ecm/webui/component/explorer/documents/UINewDocumentForm.java]: Not necessary
```suggestion
```

File: core/webui-explorer/src/main/java/org/exoplatform/ecm/webui/component/explorer/documents/UINewDocumentForm.java [hunk 4]
```diff
@@ -83,10 +86,11 @@
    * Constructor.
    *
    */
-  public UINewDocumentForm() {
+  public UINewDocumentForm() throws Exception{
     this.documentService = this.getApplicationComponent(DocumentService.class);
     // Title textbox
     UIFormStringInput titleTextBox = new UIFormStringInput(FIELD_TITLE_TEXT_BOX, FIELD_TITLE_TEXT_BOX, null);
+    titleTextBox.addValidator(MandatoryValidator.class).addValidator(XSSValidator.class);
```
Comment (rdenarie) [inline core/webui-explorer/src/main/java/org/exoplatform/ecm/webui/component/explorer/documents/UINewDocumentForm.java]: I think the mandatory validator is not necessary
Before your fix, if you leave the title empty, there is already a message. So please remove this validator (dont forget to remove the import)

File: core/webui/src/main/java/org/exoplatform/ecm/webui/form/validator/XSSValidator.java [hunk 5]
```diff
@@ -19,23 +19,33 @@
...
   public void validate(UIFormInput uiInput) throws Exception {
     String inputValue = ((String) uiInput.getValue());
     if (inputValue == null || inputValue.trim().length() == 0) {
       return;
     }
+    try{
+      inputValue = URLDecoder.decode(inputValue, StandardCharsets.UTF_8);
+    }catch (Exception e){
+      LOG.warn(e.getMessage());
+    }
```
Comment (rdenarie) [inline core/webui/src/main/java/org/exoplatform/ecm/webui/form/validator/XSSValidator.java]: The function already throws Exception, no need to catch and log it

**Review by rdenarie — CHANGES_REQUESTED** [review]

File: core/webui-explorer/src/main/java/org/exoplatform/ecm/webui/component/explorer/documents/UINewDocumentForm.java [hunk 6]
```diff
@@ -31,9 +31,7 @@
 import org.apache.commons.lang.StringUtils;
 
 import org.exoplatform.ecm.utils.text.Text;
-import org.exoplatform.ecm.webui.component.explorer.UIJCRExplorer;
-import org.exoplatform.ecm.webui.form.validator.ECMNameValidator;
-import org.exoplatform.ecm.webui.form.validator.XSSValidator;
+import org.exoplatform.ecm.webui.component.explorer.UIJCRExplorer;import org.exoplatform.ecm.webui.form.validator.XSSValidator;
```
Comment (rdenarie) [inline core/webui-explorer/src/main/java/org/exoplatform/ecm/webui/component/explorer/documents/UINewDocumentForm.java]: Please use the correct formatting

File: core/webui-explorer/src/main/java/org/exoplatform/ecm/webui/component/explorer/documents/UINewDocumentForm.java (hunk 1 above)
Comment (rdenarie) [reply core/webui-explorer/src/main/java/org/exoplatform/ecm/webui/component/explorer/documents/UINewDocumentForm.java:87]: Please fix this point

File: core/webui-explorer/src/main/java/org/exoplatform/ecm/webui/component/explorer/documents/UINewDocumentForm.java [hunk 7]
```diff
@@ -136,7 +137,6 @@ public void execute(Event<UINewDocumentForm> event) throws Exception {
       }
       // Get title
       String title = uiDocumentForm.getUIStringInput(FIELD_TITLE_TEXT_BOX).getValue();
-
```
Comment (rdenarie) [inline core/webui-explorer/src/main/java/org/exoplatform/ecm/webui/component/explorer/documents/UINewDocumentForm.java]: remove this modification

File: core/webui/src/main/java/org/exoplatform/ecm/webui/form/validator/XSSValidator.java [hunk 8]
```diff
@@ -19,24 +19,29 @@
 import org.apache.commons.lang.StringUtils;
 
 import org.exoplatform.commons.utils.HTMLSanitizer;
+import org.exoplatform.services.log.ExoLogger;
+import org.exoplatform.services.log.Log;
```
Comment (rdenarie) [inline core/webui/src/main/java/org/exoplatform/ecm/webui/form/validator/XSSValidator.java]: Not needed

File: core/webui/src/main/java/org/exoplatform/ecm/webui/form/validator/XSSValidator.java [hunk 9]
```diff
@@ -19,24 +19,29 @@
...
+import java.net.URLDecoder;
+import java.nio.charset.StandardCharsets;
+
 /**
  * Created by The eXo Platform SAS Author : Tran Hung Phong
  * phongth@exoplatform.com Jul 24, 2012
  */
 public class XSSValidator implements Validator {
-
+  private static final Log LOG = ExoLogger.getLogger(XSSValidator.class);
```
Comment (rdenarie) [inline core/webui/src/main/java/org/exoplatform/ecm/webui/form/validator/XSSValidator.java]: not needed

**Review by rdenarie — CHANGES_REQUESTED** [review]

File: core/webui-explorer/src/main/java/org/exoplatform/ecm/webui/component/explorer/documents/UINewDocumentForm.java [hunk 10]
```diff
@@ -29,9 +29,9 @@
 import javax.jcr.nodetype.NodeDefinition;
 
 import org.apache.commons.lang.StringUtils;
-
```
Comment (rdenarie) [inline core/webui-explorer/src/main/java/org/exoplatform/ecm/webui/component/explorer/documents/UINewDocumentForm.java:32]: Please remove this modification. We need to keep the minimum modified lines

**Review by rdenarie — CHANGES_REQUESTED** [review]

File: core/webui/src/main/java/org/exoplatform/ecm/webui/form/validator/XSSValidator.java [hunk 11]
```diff
@@ -24,19 +24,21 @@
...
 
+import java.net.URLDecoder;
+import java.nio.charset.StandardCharsets;
+
 /**
  * Created by The eXo Platform SAS Author : Tran Hung Phong
  * phongth@exoplatform.com Jul 24, 2012
  */
 public class XSSValidator implements Validator {
-
```
Comment (rdenarie) [inline core/webui/src/main/java/org/exoplatform/ecm/webui/form/validator/XSSValidator.java]: Please remove this modification

**Review by rdenarie — CHANGES_REQUESTED** [review]

**Review by rdenarie — APPROVED** [review]
